crates/nargo/src/cli/verify_cmd.rs

```diff
@@ -67,36 +81,42 @@ pub fn verify_with_path<P: AsRef<Path>>(
     show_ssa: bool,
 ) -> Result<bool, CliError> {
     let compiled_program = compile_circuit(program_dir.as_ref(), show_ssa)?;
+    let mut public_inputs = BTreeMap::new();
 
+    // Load public inputs (if any) from `VERIFIER_INPUT_FILE`.
     let public_abi = compiled_program.abi.clone().unwrap().public_abi();
     let num_pub_params = public_abi.num_parameters();
-    let mut public_inputs = BTreeMap::new();
     if num_pub_params != 0 {
         let curr_dir = program_dir;
         public_inputs = noirc_abi::input_parser::Format::Toml
             .parse(curr_dir, VERIFIER_INPUT_FILE)
             .map_err(CliError::from)?;
     }
 
-    if num_pub_params != public_inputs.len() {
-        // return Err(CliError::Generic(format!("")));
-        panic!(
-            "Expected {} number of values in {}.toml, but got {} number of values",
-            num_pub_params,
-            VERIFIER_INPUT_FILE,
-            public_inputs.len()
-        )
-    }
+    let valid_proof = verify_proof(compiled_program, public_inputs, load_proof(proof_path)?)?;
 
-    let public_inputs = process_abi_with_verifier_input(public_abi, public_inputs)?;
+    Ok(valid_proof)
+}
 
-    // XXX: Instead of unwrap, return a PathNotValidError
-    let proof_hex: Vec<_> = std::fs::read(&proof_path).unwrap();
-    // XXX: Instead of unwrap, return a ProofNotValidError
-    let proof = hex::decode(proof_hex).unwrap();
+fn verify_proof(
+    compiled_program: CompiledProgram,
+    public_inputs: BTreeMap<String, InputValue>,
+    proof: Vec<u8>,
+) -> Result<bool, CliError> {
+    let public_inputs =
+        process_abi_with_verifier_input(compiled_program.abi.unwrap(), public_inputs)?;
 
     let backend = crate::backends::ConcreteBackend;
     let valid_proof = backend.verify_from_cs(&proof, public_inputs, compiled_program.circuit);
 
     Ok(valid_proof)
 }
+
+fn load_proof<P: AsRef<Path>>(proof_path: P) -> Result<Vec<u8>, CliError> {
+    // XXX: Instead of unwrap, return a PathNotValidError
+    let proof_hex: Vec<_> = std::fs::read(&proof_path).unwrap();
+    // XXX: Instead of unwrap, return a ProofNotValidError
```

Comment:
Can these TODOs be added as part of this PR? I'm not the biggest fan of unwraps that can fail

Reply:
Sure, I can handle this.

Reply:
I've taken the opportunity to rework errors a bit in general. Rather than just holding plain error strings, errors hold the relevant data and the message is formatted in `errors.rs`.